Mark case copies as "Copied Case" and remove these cases from deactivated mobile users

Charl1996 opened the pull request:
This work needs to go into the base branch before the base is merged to master.

Only my own commits are relevant.

## Product Description
This PR adds a label to all cases which are copies. See the screenshot below of the Case List for example.

![image](https://github.com/dimagi/commcare-hq/assets/44245062/678ab08f-2fe8-4bda-8748-5f1c0e82eaf8)

When a case is created which is a copy of another case, a new case property will be added to the case, called `commcare_case_copy`. The mere presence of the case property signifies that this case is a copy, so the value of this new case property is less important. 
Note that `commcare_case_copy` is also a reserved case property.  

When a mobile user is deactivated, all copied/test cases will be removed from that user.

## Technical Summary
A copied/test case will now have a new case property, `commcare_case_copy`, which marks this case a copy. 

When a mobile worker is deactivated, all the copied cases assigned to this mobile worker will be removed. This is achieved by running a CaseSearchES query to get all the cases with this particular case property present belonging to this particular owner.

This removal process is triggered by the `commcare_user_post_save` django signal.

## Feature Flag
Case Copy

## Safety Assurance

### Safety story
Tested locally as well as through unit testing. This feature will also be tested on staging.

### Automated test coverage
Added tests for the new work.

### QA Plan
QA will be done manually for this feature.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Charl1996 committed:
Fetch case geo-data property from configuration

kaapstorm committed:
♻ `CaseHelper.copy_cases()` to `api._copy_cases()`

nospame committed:
Don't interpolate dots for search property validations

nospame committed:
Test interpolate_dots flag for interpolate_xpath

nospame committed:
Rename _validate_case_reference => _ensure_no_case_references

kaapstorm committed:
Drop `count_only` param

kaapstorm committed:
Add tests

kaapstorm committed:
Drop some parentheses

kaapstorm committed:
Use const instead of `function.__name__`

Rather use the const that already exists, than use `function.__name__`
as a const. `DEID_TRANSFORM_FUNCTIONS` shows that these consts are
already being used to look up the transform functions. We should either
stick with that, or we should change the current usage to use
`function.__name__` instead, and the second option looks impractical.

kaapstorm committed:
Drop properties that return consts

Now that we are just using consts, these properties aren't adding much
value.

kaapstorm committed:
Give `k`, `v` more explanatory names

kaapstorm committed:
Re-attach comment to its code

kaapstorm committed:
Apply style guide to type hints, rst to docstrings

* [Style guide](https://docs.google.com/document/d/1q90SaGuxV9dKnjnEIDdu8ma7xrZJxJq_HsfkCUhQqqk/edit)
* [reStructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html)

kaapstorm committed:
Drop type hints

kaapstorm committed:
Squash class hierarchy

`_Importer` is never used. YAGNI

kaapstorm committed:
Two ways to import exceptions --> one way

kaapstorm committed:
`do_import.py`: Add linebreaks and docstrings

kaapstorm committed:
`do_import.py`: Add type hints

kaapstorm committed:
Add `add_inferred_props_to_schema` param

I chose this approach to perserve current behavior, but I don't think
it's the right one. Instead of using a `soft_assert` in
`submit_and_process_caseblocks()` shouldn't we just check whether
`case_type` is set, and if not, don't try to add properties to its
schema?

kaapstorm committed:
`copy_cases()` uses `SubmitCaseBlockHandler`

kaapstorm committed:
Refactor `_copy_cases()` to `CaseCopier` class

* Changes `original_cases` to a dict to make retrieving cases faster.
* Uses IndexAttrs to make it clearer when an index is a dict and when
  it is a tuple.
* Adds some tests for `CaseCopier` methods

kaapstorm committed:
Drop type hints

Charl1996 committed:
Fetch geo case prop from util function

Charl1996 committed:
Merge branch 'master' into cs/fetch-geo-data-from-settings

Charl1996 committed:
Add property to indicate case is a copy

Charl1996 committed:
Restrict custom prop name to be same as copy case prop

snopoke committed:
move markdown to its own file

snopoke committed:
remove unused var

Charl1996 committed:
Check for reserved case prop on data dictionary property save

snopoke committed:
move all markdown setup to markdown.js

snopoke committed:
modernize

snopoke committed:
lazy init

snopoke committed:
rename function

Charl1996 committed:
Label copied cases as such on UI

snopoke committed:
initial markdown tests

snopoke committed:
newline rendering tests

snopoke committed:
test integrations

snopoke committed:
tests for hmac and gaen listener registration

Charl1996 committed:
Remove copied cases on couch user deactivation

Charl1996 committed:
Use CaseHelper to close cases

Charl1996 committed:
Amend tests

Charl1996 committed:
Add more tests

Charl1996 committed:
Merge branch 'cs/SC-2636-copy_cases_to_test_user' into cs/SC-2904-mark-test-cases

Charl1996 committed:
Check for mobile worker before invoking celery task

mkangia committed:
updates to use expected key

snopoke committed:
Merge branch 'master' into sk/web-apps-markdown

snopoke committed:
add cloudcareUtils back

mkangia committed:
Merge pull request #33339 from dimagi/mk/expression-debugger-show-error

Fix error message display on expression debugger

zandre-eng commented on corehq/apps/reports/filters/api.py:
```diff
@@ -355,7 +357,10 @@ def _create_caseblock(self, case):
             case_type=case.type,
-            update={**case.case_json, **deid_props},
+            update={
+                self.COMMCARE_CASE_COPY_PROPERTY_NAME: True,
```
Does it matter what the value of this property is? If not,  what do you think about having the orignal case ID as the value instead? It could be helpful in scenarios where there are problems with the copied case data. If we have the orignal ID we can then easily track it back to the original case for comparison.

Charl1996 replied:
Maybe not a bad idea

zandre-eng commented on corehq/apps/users/tasks.py:
```diff
@@ -394,10 +394,8 @@ def apply_correct_demo_mode_to_loadtest_user(commcare_user_id):
-            case_ids=test_case_ids,
-        )
+    for case_id in test_case_ids:
+        CaseHelper(domain=domain, case_id=case_id).close()
```
What was the motivation for switching it from a hard delete to closing the cases?

Charl1996 replied:
Updated this to delete again instead of closing.

snopoke committed:
Merge pull request #33332 from dimagi/sk/web-apps-markdown

refactor web apps markdown

snopoke committed:
auto-fix issues

snopoke committed:
fix simple issues

snopoke committed:
don't re-use var names

snopoke committed:
remove unused variable

snopoke committed:
use truthy to check for undefined or null

millerdev committed:
Remove unused open file

millerdev committed:
Pass path to ZipFile, eliminate superfluous open

millerdev committed:
Generate unique filenames for zip file

nospame committed:
Merge pull request #33302 from dimagi/em/search-property-dot-notation

Allow dot notation in case search property validations

millerdev committed:
Merge pull request #33343 from dimagi/dm/multimedia-filenames

Unique multimedia filenames in export zips

esoergel committed:
Sort case API by copy of ID for performance reasons

I opted to manipulate the query this way rather than integrate it into
the ESQuery framework as this is a one-off hack and not something we're
likely to be repeating.  Open to second guessing that decision, though.

This sort syntax breaks in ES6, I believe, but this code is covered by
tests, so any breaking changes should surface easily

Charl1996 committed:
Merge pull request #33268 from dimagi/cs/fetch-geo-data-from-settings

Fetch case geo-data property from configuration

snopoke committed:
Merge pull request #33341 from dimagi/sk/cloudcare-lint

Lint for cloudcare JS

esoergel committed:
Merge pull request #33349 from dimagi/es/sort-case-api

Sort case API by copy of ID for performance reasons

Charl1996 committed:
Remove test cases on deactivate

Charl1996 committed:
Merge branch 'master' into cs/SC-2904-mark-test-cases

Charl1996 committed:
Merge branch 'cs/SC-2636-copy_cases_to_test_user' into cs/SC-2904-mark-test-cases

Charl1996 commented:
Nah, something's weird now with the commits. Closing this for now.

Charl1996 commented on corehq/apps/reports/const.py:
```diff
@@ -14,3 +14,4 @@
 )
 
 HQ_TABLEAU_GROUP_NAME = 'HQ'
+COMMCARE_CASE_COPY_PROPERTY_NAME = "commcare_case_copy"
```
Yeah, I'll update
